classes/listener/observer.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace block_vmchecker\listener;
+
+defined('MOODLE_INTERNAL') || die();
+
+class observer {
+    private static function stopPreviousAttemp($assignmentId) {
+        global $DB, $USER, $CFG;
+        
+        $previousAttempt = $DB->get_record('block_vmchecker_submissions',
+            array(
+                'userid' => $USER->id,
+                'assignid' => $assignmentId,
+        ));
+        if (!$previousAttempt)
+            return;
+
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_URL, $CFG->block_vmchecker_backend . $previousAttempt->uuid . '/cancel');
+        curl_setopt($ch, CURLOPT_HTTPHEADER, array("Content-Type: application/json"));
+        curl_setopt($ch, CURLOPT_POST, true);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+
+        curl_exec($ch);
+
+        curl_close($ch);
+
+        $DB->delete_records('block_vmchecker_submissions',
+            array('id' => $previousAttempt->id)
+        );
+
+    }
+
+    public static function submit(\core\event\base $event) {
+        global $DB, $USER, $CFG;
+
+        $data = $event->get_data();
+        $submission = $DB->get_record($data['objecttable'],
+            array(
+                'id' => $data['objectid']
+        ));
+
+        $vmchecker_options = $DB->get_record('block_vmchecker_options',
+            array(
+                'assignid' => $submission->assignment
+        ));
+
+        // The assignment is not a vmchecker type
+        if ($vmchecker_options == null)
+            return;
+
+        \block_vmchecker\listener\observer::stopPreviousAttemp($submission->assignment);
+
+        $config_data = $DB->get_record('block_instances',
+            array(
+                'id' => $vmchecker_options->blockinstanceid
+        ), 'configdata')->configdata;
+        $config = unserialize(base64_decode($config_data));
+
+        $fs = get_file_storage();
+        $files = $fs->get_area_files($data['contextid'], 'assignsubmission_file', 'submission_files', $submission->submission);
+        $submited_file = null;
+        foreach($files as $file) {
+            if ($file->get_filename() != '.') {
+                $submited_file = $file;
+                break;
+            }
+        }
+
+        if (!$submited_file)
+            return;
+
+        $payload = json_encode(array(
+            'gitlab_private_token' => $config->gitlab_private_token,
+            'gitlab_project_id' => $config->gitlab_project_id,
+            'username' => $USER->username,
```

Comment:
The block has a functionality that allows the teacher to recheck a certain submission ([recheck_task.php ](https://github.com/Jokeswar/vmchecker-next/blob/master/classes/task/recheck_task.php)- does the same thing as this class, but makes sure that the uploading user is sent to the CI - duplicate  code, will fix).

Reply:
I understand that there is that functionality, but I was thinking more of what would happen if an user would do it regardless, since you cannot block them.